### DISCUSSION TITLE
Loading speed issue

### DISCUSSION THREAD

**Discussion by miracuruzrh** [JanDeDobbeleer/oh-my-posh#5440] [Troubleshoot] [answered]
Hi I'm running the following versions:
PowerShell 7.4.4
oh-my-posh 23.6.2
Theme: gruvbox

oh-my-posh is installed via the winget command.

Recently, I have recognized that the loading of my profiles are taking relatively long when I have oh-my-posh activated.
I have read [here](https://ohmyposh.dev/docs/installation/prompt) on the documentation page that the antivirus software could be the issue. and that there is an alternative command. Therefore I've tested out both variants.

I have measured following times:
Standard Variant: ~5300ms
Antivirus Variant: ~6800ms
Without oh-my-posh: ~1300ms

I use oh-my-posh already since long time.
I think, in the past it was quicker with loading the themes.

But I'm not 100% sure. Therefore I'll ask, if the loading time are okay, or if something can be adjusted, so that the profile is loading faster?

Please let me know, if additional information are required.

p.s. I've made a little script, which helped me with testing and going back and forth with the different variants.
I provide it here additionally:

```pwsh
# Initializing oh-my-posh
# Commands:
# Display Themes: Get-PoshThemes

# Options
$measureOption  = "On" # Set to on to measure the oh-my-posh startup time
$commandVariant = "A"  # Set to A for BitDefender workaround variant
                       # Set to B for normal variant
$theme          = "gruvbox"

Write-Output ("oh-my-posh " + (oh-my-posh --version))

if ($commandVariant -eq "A") {
    if ($measureOption -eq "On") {
        measure-Command { & ([ScriptBlock]::Create((oh-my-posh init pwsh --config "$env:POSH_THEMES_PATH\$theme.omp.json" --print) -join "`n")) }
    } else {
        & ([ScriptBlock]::Create((oh-my-posh init pwsh --config "$env:POSH_THEMES_PATH\$theme.omp.json" --print) -join "`n"))
    }
} elseif ($commandVariant -eq "B"){
    if ($measureOption -eq "On") {
        measure-Command { oh-my-posh init pwsh --config "$env:POSH_THEMES_PATH\$theme.omp.json" | Invoke-Expression }
    } else {
        oh-my-posh init pwsh --config "$env:POSH_THEMES_PATH\$theme.omp.json" | Invoke-Expression
    }
} else {
    Write-Output "Without oh-my-posh"
}
```

If additional information is required, please let me know.

Best regards.

**Comment by JanDeDobbeleer** [comment]
@miracuruzrh 1 second isn't normal, it shouldn't take that long. Unless something PowerShell side is taking that long, but oh-my-posh itself is loading the script in tens of milliseconds, not seconds. Rather than exclusions, did you also try disabling Defender entirely to see of the issue persists? As a test.

**Comment by miracuruzrh** [comment, marked as answer]
Hi @JanDeDobbeleer,

thanks for your reply. Yes, I did disable Defender completely for testing, but it made no difference. After spending hours trying to find the issue, I had nearly given up. However, I discovered the problem by pure luck.

It turned out that remnants of an old Wazuh-agent installation were causing the issue. I had briefly tested Wazuh some time ago but never really continued with it. Although the server was no longer there, the agent was still installed. As I was stuck with the oh-my-posh configuration issue, I decided to revisit Wazuh. I set up a new server and reinstalled the agent on the computer with the slow oh-my-posh configuration. Since then, my load times have improved significantly. Using method A (BitDefender workaround), I now have an average startup time of 270 ms. With method B (standard variant), it's around 310 ms. These times are much better. If I disable the speed measuring, it feels even faster, as the measurement itself takes some time as well. Would you say these values are normal, or still place for improvement?

I've also adjusted my script in case anyone is interested (it shows now only one line with the milliseconds):
```powershell
# Initializing oh-my-posh
# Commands:
# Display Themes: Get-PoshThemes

# Options
$measureOption  = "On" # Set to on to measure the oh-my-posh startup time
$commandVariant = "A"  # Set to A for BitDefender workaround variant
                       # Set to B for normal variant
$theme          = "gruvbox"

Write-Output ("oh-my-posh " + (oh-my-posh --version))

if ($commandVariant -eq "A") {
    if ($measureOption -eq "On") {
        $time = Measure-Command { & ([ScriptBlock]::Create((oh-my-posh init pwsh --config "$env:POSH_THEMES_PATH\$theme.omp.json" --print) -join "`n")) }
        $formattedTime = [math]::Round($time.TotalMilliseconds)
        Write-Output ("Startup time: " + $formattedTime + " ms")
    } else {
        & ([ScriptBlock]::Create((oh-my-posh init pwsh --config "$env:POSH_THEMES_PATH\$theme.omp.json" --print) -join "`n"))
    }
} elseif ($commandVariant -eq "B"){
    if ($measureOption -eq "On") {
        $time = Measure-Command { oh-my-posh init pwsh --config "$env:POSH_THEMES_PATH\$theme.omp.json" | Invoke-Expression }
        $formattedTime = [math]::Round($time.TotalMilliseconds)
        Write-Output ("Startup time: " + $formattedTime + " ms")
    } else {
        oh-my-posh init pwsh --config "$env:POSH_THEMES_PATH\$theme.omp.json" | Invoke-Expression
    }
} else {
    Write-Output "Without oh-my-posh"
}
```

**Reply by JanDeDobbeleer** [reply]
@miracuruzrh That's normal. The duration is now PowerShell's `Invoke-Expression` being slower than eval in other shells.

**Reply by miracuruzrh** [reply]
Alright. Thank you.
